src/components/search/Filter/useFilter.ts

```diff
@@ -0,0 +1,118 @@
+import type { IStoreSelectedFacet } from '@faststore/api'
+import { useSearch } from '@faststore/sdk'
+import type { Filter_FacetsFragment } from '@generated/graphql'
+import { useReducer, useMemo, useEffect } from 'react'
+
+interface State {
+  expanded: Set<number>
+  selected: IStoreSelectedFacet[]
+}
+
+type Action =
+  | {
+      type: 'toggleExpanded'
+      payload: number
+    }
+  | {
+      type: 'selectFacets'
+      payload: IStoreSelectedFacet[]
+    }
+  | {
+      type: 'toggleFacet'
+      payload: IStoreSelectedFacet
+    }
+
+const reducer = (state: State, action: Action) => {
+  const { expanded, selected } = state
+  const { type, payload } = action
+
+  if (type === 'toggleExpanded') {
+    if (expanded.has(payload)) {
+      expanded.delete(payload)
+    } else {
+      expanded.add(payload)
+    }
+
+    return {
+      ...state,
+      expanded: new Set(expanded),
+    }
+  }
+
+  if (type === 'selectFacets' && payload !== selected) {
+    return {
+      ...state,
+      selected: payload,
+    }
+  }
+
+  if (type === 'toggleFacet') {
+    const index = state.selected.findIndex(
+      (facet) => facet.key === payload.key && facet.value === payload.value
+    )
+
+    if (index > -1) {
+      return {
+        ...state,
+        selected: state.selected.filter((_, idx) => idx !== index),
+      }
+    }
+
+    return {
+      ...state,
+      selected: [...state.selected, payload],
+    }
+  }
+
+  return state
+}
+
+export const useFilter = (allFacets: Filter_FacetsFragment[]) => {
+  const {
+    state: { selectedFacets },
+  } = useSearch()
+
+  const [{ selected, expanded }, dispatch] = useReducer(reducer, null, () => ({
+    expanded: new Set([]),
+    selected: selectedFacets,
+  }))
+
+  const selectedMap = useMemo(
```

Comment:
Using a HasMap isn't better than an object?

Reply:
yes, I'll change it

Reply:
done